[Lens as code] Add Heatmap schema and transforms

nickofthyme opened the pull request:
## Summary

Adds API schemas and transforms for **Heatmap**

Closes #234575 
Closes #234599

### Checklist

- [x] The PR  description includes the appropriate Release Notes section, and the correct `release_note:*` label is applied per the [guidelines](https://www.elastic.co/guide/en/kibana/master/contributing.html#kibana-release-notes-process)
- [x] Review the [backport guidelines](https://docs.google.com/document/d/1VyN5k91e5OVumlc0Gb9RPa3h1ewuPE705nRtioPiTvY/edit?usp=sharing) and apply applicable `backport:*` labels.

nickofthyme committed:
feat: add heatmap api schemas

nickofthyme commented on src/platform/packages/shared/kbn-lens-embeddable-utils/config_builder/schema/charts/heatmap.ts:
```diff
@@ -0,0 +1,118 @@
+      ],
+      { defaultValue: 'auto' }
+    )
+  ),
```
Could we align this with...

https://github.com/elastic/kibana/blob/c152f868f796664103230409cfaba325ae4c7b59/src/platform/packages/shared/kbn-lens-embeddable-utils/config_builder/schema/charts/legacy_metric.ts#L29-L41

nickofthyme committed:
chore: simply common layer logic into `getDataSourceLayer`

nickofthyme committed:
chore: update heatmap schema

nickofthyme committed:
chore: split dataset util functions

nickofthyme committed:
feat: add heatmap api transform logic

nickofthyme committed:
tests: add heatmap SO integration tests

nickofthyme committed:
Merge branch 'main' into heatmap-schema

nickofthyme committed:
Merge branch 'main' into heatmap-schema

nickofthyme committed:
Merge branch 'main' into heatmap-schema

nickofthyme committed:
fix: missing schema `query` property

nickofthyme committed:
fix: lazy builder import failing on load

nickofthyme committed:
fix: heatmap missing configs

- split up heatmap transform to/from api
 - missing title configs
 - missing rotation configs
 - missing cell label visibility config even in schema

nickofthyme committed:
Merge branch 'main' into heatmap-schema

nickofthyme committed:
Merge branch 'main' into heatmap-schema

nickofthyme committed:
fix: merge issues

nickofthyme committed:
test: add color test mock SO

nickofthyme committed:
fix: lazy builder types

nickofthyme committed:
Merge branch 'main' into heatmap-schema

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/368047)
* Commit: a7a988781f25dcb608ede446cbe22eb27a257ed2

### Failed CI Steps
* [Jest Tests #20](https://buildkite.com/elastic/kibana-pull-request/builds/368047#019ae20f-e031-40a9-90b3-d4d4c26204b2)
* [Jest Tests #20](https://buildkite.com/elastic/kibana-pull-request/builds/368047#019ae225-ef6f-49aa-830f-e536cdfc0909)

### Test Failures
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/368047#019ae225-ef6f-49aa-830f-e536cdfc0909) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/368047/jobs/019ae225-ef6f-49aa-830f-e536cdfc0909/artifacts/019ae24c-d608-4a9a-99a6-360aadec527b) Jest Tests #<span></span>20 / <AboutSection /> should render about section for signal document
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/368047#019ae20f-e031-40a9-90b3-d4d4c26204b2) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/368047/jobs/019ae20f-e031-40a9-90b3-d4d4c26204b2/artifacts/019ae225-d7e9-4030-9cef-467d18821c71) Jest Tests #<span></span>20 / buildDatasetState builds index dataset state
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/368047#019ae225-ef6f-49aa-830f-e536cdfc0909) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/368047/jobs/019ae225-ef6f-49aa-830f-e536cdfc0909/artifacts/019ae24c-d609-407d-8df8-ce3ff861de21) Jest Tests #<span></span>20 / buildDatasetState builds index dataset state

### History
* :broken_heart: [Build #367948](https://buildkite.com/elastic/kibana-pull-request/builds/367948) failed cdebe676e10508c47eaa21db09d4aa9729db8879
* :broken_heart: [Build #367434](https://buildkite.com/elastic/kibana-pull-request/builds/367434) failed e578afe2a4abc3550d189f5cc3df6b3473b5424c

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/368047","number":368047,"commit":"a7a988781f25dcb608ede446cbe22eb27a257ed2"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/367948","number":367948,"commit":"cdebe676e10508c47eaa21db09d4aa9729db8879"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/367434","number":367434,"commit":"e578afe2a4abc3550d189f5cc3df6b3473b5424c"}],"number":368047}
buildkite-pr-comment-->

andrimal commented on src/platform/packages/shared/kbn-lens-embeddable-utils/config_builder/transforms/charts/gauge.ts:
```diff
@@ -258,7 +258,7 @@ export function fromAPItoLensState(config: GaugeState): GaugeAttributesWithoutFi
 export function fromLensStateToAPI(
   config: LensAttributes
-): Extract<LensApiState, { type: 'gauge' }> {
+): Extract<LensApiState, { type?: 'gauge' }> {
```
Is this change needed?

nickofthyme replied:
You're right, I think this was causing issues before but looks fine now as it was. Reverted in f9632e2

nickofthyme committed:
chore: add better meta data for new heatmap schemas

nickofthyme committed:
chore: cleanup duplicate `orientation` and `title` schemas

nickofthyme committed:
fix: cell labels schema and defaults

nickofthyme committed:
fix: use `DEFAULT_LAYER_ID` whenever possible

nickofthyme committed:
chore: cleanup constants

nickofthyme committed:
chore: revert gauge type change

nickofthyme committed:
cleanup: query and filter types

nickofthyme commented on src/platform/packages/shared/kbn-lens-embeddable-utils/config_builder/transforms/utils.ts:
```diff
@@ -225,7 +218,27 @@ export const buildDatasetState = (
+  }
+
+  return buildDatasetStateNoESQL(layer, layerId, adHocDataViews, references, adhocReferences);
+}
```
Using this `buildDatasetState` function creates type issues as previously we call this before knowing is the config is ES|QL or not, thus `dataset` was always either type and force casted.

Now we should avoid using `buildDatasetState` and instead use `buildDatasetStateESQL` or `buildDatasetStateNoESQL` where needed, makes the types way easier to work with and no need for additional type guards.

nickofthyme committed:
Merge remote-tracking branch 'origin/heatmap-schema' into heatmap-schema

nickofthyme committed:
Merge branch 'main' into heatmap-schema

nickofthyme commented on src/platform/packages/shared/kbn-lens-embeddable-utils/config_builder/transforms/charts/heatmap/constants.ts:
```diff
@@ -0,0 +1,14 @@
+export const API_VISUALIZATION_TYPE = 'heatmap';
+export const ACCESSOR = 'heatmap_value_accessor';
+export const LENS_DEFAULT_LAYER_ID = 'layer_0';
+export const HEATMAP_GRID_NAME = 'heatmap_grid';
```
Good point, thanks. I removed this file and just used the existing constants. See 93c27d0.

It would be nice to improve these a bit when the dust settles, seems like constants and constant functions (like the accessors) are all over the place.

nickofthyme commented on src/platform/packages/shared/kbn-lens-embeddable-utils/config_builder/schema/charts/heatmap.ts:
```diff
@@ -0,0 +1,129 @@
+  ),
+  cells: schema.maybe(
+    schema.object({
+      labels: schema.maybe(simpleLabelsSchema),
```
Yeah good point, it's hard to share schemas that look the same but might have alternate defaults or descriptions. I just created a new one for the `cells.labels` schema, see 43cada8.

nickofthyme commented on src/platform/packages/shared/kbn-lens-embeddable-utils/config_builder/schema/charts/heatmap.ts:
```diff
@@ -0,0 +1,129 @@
+      [schema.literal('horizontal'), schema.literal('vertical'), schema.literal('angled')],
+      { defaultValue: 'horizontal' }
+    )
+  ),
```
Yup, it seems there are a few duplications of simple enums like this. I think I'd be best to store all these in one place and make it easier to discover, something like below. See more details in 266e0fd.

```ts
  orientation: schema.maybe(
    builderEnums.orientation({
      defaultValue: 'horizontal',
      meta: { description: 'Orientation of the axis labels' },
    })
  ),
```

nickofthyme replied:
Yup exactly, I just left them for now

nickofthyme committed:
chore: remove constant export

nickofthyme committed:
fix: schema utils and cleanup shared props

nickofthyme commented on src/platform/packages/shared/kbn-lens-embeddable-utils/config_builder/schema/charts/heatmap.ts:
```diff
@@ -0,0 +1,129 @@
+  ...heatmapAxesStateSchema,
+  ...dslOnlyPanelInfoSchema,
+  ...datasetSchema,
+  metric: mergeAllMetricsWithChartDimensionSchema(
```
I'll take your word on it 😄, fixed in 967a2a7.

nickofthyme committed:
Merge branch 'main' into heatmap-schema

nickofthyme commented on src/platform/packages/shared/kbn-lens-embeddable-utils/config_builder/schema/index.ts:
```diff
@@ -48,20 +51,28 @@ export const _lensApiStateSchema: any = schema.oneOf([
+  | MetricState
+  | LegacyMetricState
+  | GaugeState
+  | XYState
```
The missing `XYState` was causing many of the `@ts-expect-error` issues removed in this pr.

nickofthyme committed:
Merge branch 'main' into heatmap-schema

nickofthyme commented on src/platform/packages/shared/kbn-lens-embeddable-utils/config_builder/config_builder.ts:
Another issue was with the `isVisible`.

```diff
"legend": Object {
   "position": "right",
-  "visible": true,
+  "visible": undefined,
},
```

...which is caused by `isVisible` being `boolean` type here...

https://github.com/elastic/kibana/blob/0d270a1caf0cb10d50a37f9ca423eb3358907b7b/src/platform/packages/shared/kbn-lens-common/visualizations/heatmap/types.ts#L19-L23

...but the integrations SO actually leaved it `undefined` (missing), so in the transforms I need to default to some value if undefined. I think this can be cleaned up later.

nickofthyme replied:
The same thin happens for `isYAxisTitleVisible` and `isXAxisTitleVisible`. 😞 

```diff
      "x": Object {
        "labels": Object {
          "visible": true,
        },
        "title": Object {
          "value": undefined,
-         "visible": false,
+         "visible": undefined,
        },
      },
      "y": Object {
        "labels": Object {
          "visible": true,
        },
        "title": Object {
          "value": undefined,
-         "visible": false,
+         "visible": undefined,
        },
      },
```

andrimal replied:
Yes exactly, these `undefined` title/legend visibility values in old SOs are what I was talking about.

nickofthyme reviewed (commented):
@andrimal I looked at the integrations and found a few issues apart from the ids and default issues.

nickofthyme committed:
chore: fix throw check

- the string passed to `toThrow` is not used as an error message but instead a compare check for thrown errors.

nickofthyme committed:
fix: undefined property logic

nickofthyme committed:
fix: merge error in test

nickofthyme committed:
Merge branch 'main' into heatmap-schema

elasticmachine commented:
## :broken_heart: Build Failed
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/372117)
* Commit: 0d7906ce1c44ce854439ad2e023d32e890c34131

### Failed CI Steps
* [Check Types](https://buildkite.com/elastic/kibana-pull-request/builds/372117#019b0f42-909f-47a9-a133-6fd8a246f176)

### History
* :broken_heart: [Build #371932](https://buildkite.com/elastic/kibana-pull-request/builds/371932) failed 967a2a71a833420e117bd7b51d6564eadd6c94da
* :broken_heart: [Build #371518](https://buildkite.com/elastic/kibana-pull-request/builds/371518) failed b8de214f37d1f1f51936e45aa85c5b4c41c30349
* :broken_heart: [Build #368047](https://buildkite.com/elastic/kibana-pull-request/builds/368047) failed a7a988781f25dcb608ede446cbe22eb27a257ed2
* :broken_heart: [Build #367948](https://buildkite.com/elastic/kibana-pull-request/builds/367948) failed cdebe676e10508c47eaa21db09d4aa9729db8879
* :broken_heart: [Build #367434](https://buildkite.com/elastic/kibana-pull-request/builds/367434) failed e578afe2a4abc3550d189f5cc3df6b3473b5424c

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/372117","number":372117,"commit":"0d7906ce1c44ce854439ad2e023d32e890c34131"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/371932","number":371932,"commit":"967a2a71a833420e117bd7b51d6564eadd6c94da"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/371518","number":371518,"commit":"b8de214f37d1f1f51936e45aa85c5b4c41c30349"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/368047","number":368047,"commit":"a7a988781f25dcb608ede446cbe22eb27a257ed2"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/367948","number":367948,"commit":"cdebe676e10508c47eaa21db09d4aa9729db8879"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/367434","number":367434,"commit":"e578afe2a4abc3550d189f5cc3df6b3473b5424c"}],"number":372117}
buildkite-pr-comment-->

nickofthyme committed:
fix: bad types

nickofthyme committed:
Merge branch 'main' into heatmap-schema

elasticmachine commented:
## :yellow_heart: Build succeeded, but was flaky
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/372195)
* Commit: e8e1c728a39e753a8fd15f3832e6a16e4484d997

### Failed CI Steps
* [FTR Configs #13](https://buildkite.com/elastic/kibana-pull-request/builds/372195#019b113e-53f8-4640-8b72-944085d6564d)
* [FTR Configs #80](https://buildkite.com/elastic/kibana-pull-request/builds/372195#019b113e-544d-4bfc-85b6-9168d3e977bb)
* [FTR Configs #102](https://buildkite.com/elastic/kibana-pull-request/builds/372195#019b113e-5467-47ef-b6fa-8093594658d0)

### Test Failures
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/372195#019b113e-544d-4bfc-85b6-9168d3e977bb) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/372195/jobs/019b113e-544d-4bfc-85b6-9168d3e977bb/artifacts/019b1165-4b31-4528-8e95-28bbf15c7ffa) FTR Configs #<span></span>80 / serverless observability UI - Cases and Rules Serverless Observability Cases Cases list row actions Severity to critical
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/372195#019b113e-5467-47ef-b6fa-8093594658d0) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/372195/jobs/019b113e-5467-47ef-b6fa-8093594658d0/artifacts/019b1163-c7fd-466e-a23c-d98c0fd28c4a) FTR Configs #<span></span>102 / Synthetics API Tests SyncGlobalParamsSpaces "before all" hook for "create test private locations"
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/372195#019b113e-53f8-4640-8b72-944085d6564d) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/372195/jobs/019b113e-53f8-4640-8b72-944085d6564d/artifacts/019b1158-9761-42df-8da7-311a6b65a2a2) FTR Configs #<span></span>13 / Visualizations - Group 3 lens app - TSVB Open in Lens Gauge "before each" hook for "should show the "Convert to Lens" menu item"

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Module Count
> Fewer modules leads to a faster build time


| id | [before](https://github.com/elastic/kibana/commit/97fd89bae52a9cd05dfb026586f6f3c971520e40) | [after](https://github.com/elastic/kibana/commit/e8e1c728a39e753a8fd15f3832e6a16e4484d997) | diff |
| --- | --- | --- | --- |
| `dashboard` | 983 | 988 | +5 |
| `discover` | 1878 | 1883 | +5 |
| `infra` | 1757 | 1762 | +5 |
| `lens` | 1655 | 1660 | +5 |
| `observability` | 1651 | 1656 | +5 |
| `observabilityAIAssistantApp` | 784 | 789 | +5 |
| `onechat` | 1557 | 1562 | +5 |
| `unifiedDocViewer` | 773 | 778 | +5 |
| total |  |  | +40 |

#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/97fd89bae52a9cd05dfb026586f6f3c971520e40) | [after](https://github.com/elastic/kibana/commit/e8e1c728a39e753a8fd15f3832e6a16e4484d997) | diff |
| --- | --- | --- | --- |
| `dashboard` | 704.6KB | 708.8KB | +4.3KB |
| `discover` | 1.3MB | 1.3MB | +4.3KB |
| `infra` | 1.1MB | 1.1MB | +4.3KB |
| `lens` | 1.6MB | 1.9MB | :warning: +278.8KB |
| `observability` | 1.7MB | 1.7MB | +4.3KB |
| `observabilityAIAssistantApp` | 261.2KB | 265.5KB | +4.3KB |
| `onechat` | 477.4KB | 481.7KB | +4.3KB |
| `unifiedDocViewer` | 366.1KB | 370.4KB | +4.3KB |
| total |  |  | +308.9KB |

#### Page load bundle
> Size of the bundles that are downloaded on every page load. Target size is below 100kb


| id | [before](https://github.com/elastic/kibana/commit/97fd89bae52a9cd05dfb026586f6f3c971520e40) | [after](https://github.com/elastic/kibana/commit/e8e1c728a39e753a8fd15f3832e6a16e4484d997) | diff |
| --- | --- | --- | --- |
| `dashboard` | 18.0KB | 18.0KB | -1.0B |
| `lens` | 64.4KB | 64.5KB | +64.0B |
| total |  |  | +63.0B |

<details><summary>Unknown metric groups</summary>

#### References to deprecated APIs
| id | [before](https://github.com/elastic/kibana/commit/97fd89bae52a9cd05dfb026586f6f3c971520e40) | [after](https://github.com/elastic/kibana/commit/e8e1c728a39e753a8fd15f3832e6a16e4484d997) | diff |
| --- | --- | --- | --- |
| `@kbn/lens-embeddable-utils` | 34 | 36 | +2 |


</details>


### History
* :broken_heart: [Build #372117](https://buildkite.com/elastic/kibana-pull-request/builds/372117) failed 0d7906ce1c44ce854439ad2e023d32e890c34131
* :broken_heart: [Build #371932](https://buildkite.com/elastic/kibana-pull-request/builds/371932) failed 967a2a71a833420e117bd7b51d6564eadd6c94da
* :broken_heart: [Build #371518](https://buildkite.com/elastic/kibana-pull-request/builds/371518) failed b8de214f37d1f1f51936e45aa85c5b4c41c30349
* :broken_heart: [Build #368047](https://buildkite.com/elastic/kibana-pull-request/builds/368047) failed a7a988781f25dcb608ede446cbe22eb27a257ed2
* :broken_heart: [Build #367948](https://buildkite.com/elastic/kibana-pull-request/builds/367948) failed cdebe676e10508c47eaa21db09d4aa9729db8879
* :broken_heart: [Build #367434](https://buildkite.com/elastic/kibana-pull-request/builds/367434) failed e578afe2a4abc3550d189f5cc3df6b3473b5424c

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/372195","number":372195,"commit":"e8e1c728a39e753a8fd15f3832e6a16e4484d997"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/372117","number":372117,"commit":"0d7906ce1c44ce854439ad2e023d32e890c34131"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/371932","number":371932,"commit":"967a2a71a833420e117bd7b51d6564eadd6c94da"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/371518","number":371518,"commit":"b8de214f37d1f1f51936e45aa85c5b4c41c30349"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/368047","number":368047,"commit":"a7a988781f25dcb608ede446cbe22eb27a257ed2"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/367948","number":367948,"commit":"cdebe676e10508c47eaa21db09d4aa9729db8879"}],"number":372195}
buildkite-pr-comment-->
